src/lib.rs

```diff
@@ -411,7 +415,7 @@ impl<S: SuffixScheme> FileRotate<S> {
         path: P,
         suffix_scheme: S,
         content_limit: ContentLimit,
-        compression: Compression,
+        #[cfg(feature = "compression")] compression: Compression,
```

Comment:
Not a fan of making the API contingent upon features; it makes swapping between features in a project using this library harder because the user will need to modify code each time. I suggest we instead panic if this argument is something that's not `Compression::None` when compression is disabled.

Reply:
I agree that feature-gating args out of a parameter list is unclean. However, I'm not a fan of API that looks correct and smells correct but panics at runtime; that seems like it would just move the developer pain from compile-time to runtime and be more difficult to catch.

Some other ideas:
- Instead of panicking if an API user tries to enable compression without the feature we no-op. This is still misleading, but at least it doesn't cause unexpected panics.
- A builder pattern where we have feature-gated extension traits that add more setter functions. This would still require API users to make code changes when they enable/disable features, but it's limited to adding/removing a `use` on the extension trait. I've seen other popular libraries do this: as an arbitrary example, here's [a trait of mac-specific builder functions that winit](https://docs.rs/winit/0.29.15/x86_64-apple-darwin/winit/platform/macos/trait.WindowBuilderExtMacOS.html) only implements for that specific platform.

Also, I'd hate to make you feel pressured to merge a PR you dislike so please don't feel like you need to merge this now just to clean it up later. If there's a specific way you want to see this implemented I'm happy to alter the PR. Or if this isn't something you want to consider right now that's also fine: I'm not in any rush.

Reply:
No-op sounds like a good solution. Ideally with a warning message printed at runtime (or compile time, if possible, but no compile/runtime errors) so we don't get bug reports here and users can easily identify what's wrong.

Reply:
Let me present an argument for panicking. The panic will happen only on creation of FileRotate, and I can only really see this object being constructed in the initialization (or very early) in the application when it runs. So any simple test of the application should show this - manual or automatic integration tests.

Here is one example from a popular library, not really of a panic but of silently leaving an application non-functioning due to a programmer error: If you use `actix-web` to build a web server, and you make a mistake in a request handler's parameters asking for [app data](https://docs.rs/actix-web/latest/actix_web/web/struct.Data.html) of a type that has not been [registered](https://docs.rs/actix-web/latest/actix_web/struct.App.html#method.app_data), it will not warn you at compile time... but your API will always return some `5xx` http error saying that it's wrongly configured. Has it ever been a problem for me as a user of actix-web? No, because it makes all my tests fail and it's easily fixed.

Reply:
That said, a builder pattern sounds like a good solution as well. And it doesn't have to be through an extension trait IMO but maybe just a feature gated function on the builder? I agree that changing the number of parameters with `#[cfg]` can lead to confusion, but what do you think about feature gating builder functions, @kstrafe ?